### PR TITLE
Function description registry

### PR DESCRIPTION
**Pull request by sogasg**

### CODE REVIEW

File: Web3Semantics.FunctionDescriptionRegistry/study.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+### Initial Thoughts
+Contract Metadata (https://solidity.readthedocs.io/en/v0.5.3/metadata.html) solves the problem of having an ABI, NatSpec and source code for all contracts and there functions.
+However, a contract deployed today might only consist of a small percentage of custom functions. The rest of the functions might be common between a lot of contracts. Even if you write new functions they might be a function that somebody else has already written.
+
+// TODO: add more advantages
+
+We propose a registry for linking function signatures to their human-readable description (could, for instance, be RadSpac or NetSpac).
+
+* A HashLink would be pointing to the human-readable description swarm / IPFS / etc.
```
Comment (dOrgJelli) [inline Web3Semantics.FunctionDescriptionRegistry/study.md]: Potential rewording:
"On-chain, the human-readable description would take the form of a hash-link to a hash-addressable storage network (Swarm, IPFS, etc)."

File: Web3Semantics.FunctionDescriptionRegistry/study.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+The registry would have four functions:
+1. `addDescription(string functionSignature, string descriptionHashLink) payable`
+
+Ff there is no description for that functionSignature it will be added and the paid Ether will be set up as you stake on the description.
```
Comment (dOrgJelli) [inline Web3Semantics.FunctionDescriptionRegistry/study.md]: *If

Comment (dOrgJelli) [inline Web3Semantics.FunctionDescriptionRegistry/study.md]: as *your stake

File: Web3Semantics.FunctionDescriptionRegistry/study.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+2. `disputeDescription(string functionSignature, string newDescriptionHashLink, string functionCodeHashLink ) payable`
+
+If you want to challenge a functionSignature => functionDescription mapping you call this function with you alternativeDescription and the stake you want to provide.
```
Comment (dOrgJelli) [inline Web3Semantics.FunctionDescriptionRegistry/study.md]: with *your alternative...

File: Web3Semantics.FunctionDescriptionRegistry/study.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+The dispute will open a prediction market (Augur, Gnosis etc.). On the market, you would be able to put money on the description you believe will win. The dApp presenting the market should also link to the functions source code (from the contract metadata). If the disputer wins the market he gets the stake connected to the previous description.
+
+// Where does the stake of the disputer go if he looses? It gets added to the functions stake?
+
+3. `addStake(string functionSignature, string descriptionHashLink) payable`
+
+Lets you add a stake to a functionSignature => descriptionHashLink mapping.
```
Comment (dOrgJelli) [inline Web3Semantics.FunctionDescriptionRegistry/study.md]: The economic game you're eluding to here is still a bit fuzzy to me. Would love to see a more detailed breakdown of what you're thinking.

Comment (sogasg) [reply Web3Semantics.FunctionDescriptionRegistry/study.md]: The details are a bit fuzzy to me too. It needs more work. I hope this is good enough for the  "Initial Thoughts" section. I hope this conveys some of the big picture ideas.

I will work on it and write a more detailed description once it is prioritized

**Pull request merged**